src/Libraries/Microsoft.Extensions.AI.Abstractions/SpeechToText/SpeechToTextResponseUpdateExtensions.cs

```diff
@@ -0,0 +1,167 @@
+
+        foreach (var update in updates)
+        {
+            ProcessUpdate(update, choices, response);
+        }
+
+        AddChoicesToCompletion(choices, response, coalesceContent);
+
+        return response;
+    }
+
+    /// <summary>Combines <see cref="SpeechToTextResponseUpdate"/> instances into a single <see cref="SpeechToTextResponse"/>.</summary>
+    /// <param name="updates">The updates to be combined.</param>
+    /// <param name="coalesceContent">
+    /// <see langword="true"/> to attempt to coalesce contiguous <see cref="AIContent"/> items, where applicable,
+    /// into a single <see cref="AIContent"/>, in order to reduce the number of individual content items that are included in
+    /// the manufactured <see cref="SpeechToTextMessage"/> instances. When <see langword="false"/>, the original content items are used.
+    /// The default is <see langword="true"/>.
+    /// </param>
+    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for cancellation requests. The default is <see cref="CancellationToken.None"/>.</param>
+    /// <returns>The combined <see cref="SpeechToTextResponse"/>.</returns>
+    public static Task<SpeechToTextResponse> ToSpeechToTextResponseAsync(
+        this IAsyncEnumerable<SpeechToTextResponseUpdate> updates, bool coalesceContent = true, CancellationToken cancellationToken = default)
+    {
+        _ = Throw.IfNull(updates);
+
+        return ToResponseAsync(updates, coalesceContent, cancellationToken);
+
+        static async Task<SpeechToTextResponse> ToResponseAsync(
+            IAsyncEnumerable<SpeechToTextResponseUpdate> updates, bool coalesceContent, CancellationToken cancellationToken)
+        {
+            SpeechToTextResponse response = new([]);
+            Dictionary<int, SpeechToTextMessage> choices = [];
+
+            await foreach (var update in updates.WithCancellation(cancellationToken).ConfigureAwait(false))
+            {
+                ProcessUpdate(update, choices, response);
+            }
+
+            AddChoicesToCompletion(choices, response, coalesceContent);
+
+            return response;
+        }
+    }
+
+    /// <summary>Processes the <see cref="SpeechToTextResponseUpdate"/>, incorporating its contents into <paramref name="choices"/> and <paramref name="response"/>.</summary>
+    /// <param name="update">The update to process.</param>
+    /// <param name="choices">The dictionary mapping <see cref="SpeechToTextResponseUpdate.ChoiceIndex"/> to the <see cref="SpeechToTextMessage"/> being built for that choice.</param>
+    /// <param name="response">The <see cref="SpeechToTextResponse"/> object whose properties should be updated based on <paramref name="update"/>.</param>
+    private static void ProcessUpdate(SpeechToTextResponseUpdate update, Dictionary<int, SpeechToTextMessage> choices, SpeechToTextResponse response)
+    {
+        response.ResponseId ??= update.ResponseId;
+        response.ModelId ??= update.ModelId;
```

Comment:
My draft PR at https://github.com/dotnet/extensions/pull/5998 switches this to be on a last-wins model. I think that's a more desirable ordering, especially if multiple responses might be coming back, in which case you want this response object to more effectively represent the last rather than first.